File: iq/27/lower/find_test.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package lower_test

import (
	"testing"

	"github.com/skhal/lab/iq/27/lower"
)

func TestFind(t *testing.T) {
	tests := []struct {
		name   string
		nn     []int
		x      int
		want   int
		wantOk bool
	}{
		// size 1
		{
			name:   "size one below",
			nn:     []int{1},
			x:      0,
			want:   1,
			wantOk: true,
		},
		{
			name:   "size one match",
			nn:     []int{1},
			x:      1,
			want:   1,
			wantOk: true,
		},
		{
			name: "size one above",
			nn:   []int{1},
			x:    2,
		},
		// size 2
		{
			name:   "size two below",
			nn:     []int{1, 3},
			x:      0,
			want:   1,
			wantOk: true,
		},
		{
			name:   "size two match first",
			nn:     []int{1, 3},
			x:      1,
			want:   1,
			wantOk: true,
		},
		{
			name:   "size two below second",
			nn:     []int{1, 3},
			x:      2,
			want:   3,
			wantOk: true,
		},
		{
			name:   "size two match second",
			nn:     []int{1, 3},
			x:      3,
			want:   3,
			wantOk: true,
		},
		{
			name: "size two above",
			nn:   []int{1, 3},
			x:    4,
		},
		// size 3
		{
			name:   "size three below",
			nn:     []int{1, 3, 5},
			x:      0,
			want:   1,
			wantOk: true,
		},
		{
			name:   "size three match first",
			nn:     []int{1, 3, 5},
			x:      1,
			want:   1,
			wantOk: true,
		},
		{
			name:   "size three below second",
			nn:     []int{1, 3, 5},
			x:      2,
			want:   3,
			wantOk: true,
		},
		{
			name:   "size three match second",
			nn:     []int{1, 3, 5},
			x:      3,
			want:   3,
			wantOk: true,
		},
		{
			name:   "size three below third",
			nn:     []int{1, 3, 5},
			x:      4,
			want:   5,
			wantOk: true,
		},
		{
			name:   "size three match third",
			nn:     []int{1, 3, 5},
			x:      5,
			want:   5,
			wantOk: true,
		},
		{
			name: "size three above",
			nn:   []int{1, 3, 5},
			x:    6,
		},
	}
	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got, gotOk := lower.Find(tc.nn, tc.x)

			if gotOk != tc.wantOk {
				t.Errorf("lower.Find(%v, %d) = _, %v; want %v", tc.nn, tc.x, gotOk, tc.wantOk)
			}
			if got != tc.want {
				t.Errorf("lower.Find(%v, %d) = %d, _; want %d", tc.nn, tc.x, got, tc.want)
			}
		})
	}
}

File: iq/27/lower/find_test.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
//
// clang-format off-next-line
//go:build ignore

#include "iq/27/lower/find.h"

#include <ostream>
#include <string_view>
#include <vector>

#include "absl/status/status_matchers.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_format.h"
#include "absl/strings/str_join.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace iq::lower {
namespace {

using ::absl_testing::IsOk;
using ::absl_testing::IsOkAndHolds;
using ::testing::Not;
using ::testing::TestWithParam;
using ::testing::ValuesIn;

struct FindTestParam {
  std::string_view name;
  std::vector<int> nn;
  int k;
  int want;
};

std::ostream& operator<<(std::ostream& os, const FindTestParam& tp) {
  return os << absl::StrFormat("{ nn:[%s] k:%d }", absl::StrJoin(tp.nn, " "),
                               tp.k);
}

class FindTest : public TestWithParam<FindTestParam> {};

using PositiveFindTest = FindTest;

TEST_P(PositiveFindTest, Test) {
  const FindTestParam& tp = GetParam();

  const absl::StatusOr<int> got = Find(tp.nn, tp.k);

  EXPECT_THAT(got, IsOkAndHolds(tp.want));
}

using NegativeFindTest = FindTest;

TEST_P(NegativeFindTest, Test) {
  const FindTestParam& tp = GetParam();

  const absl::StatusOr<int> got = Find(tp.nn, tp.k);

  EXPECT_THAT(got, Not(IsOk()));
}

const FindTestParam kPositiveFindTestParams[]{
    // size 1
    {.name = "size one below", .nn = {1}, .k = 0, .want = 1},
    {.name = "size one match", .nn = {1}, .k = 1, .want = 1},
    // size 2
    {.name = "size two below", .nn = {1, 3}, .k = 0, .want = 1},
    {.name = "size two match first", .nn = {1, 3}, .k = 1, .want = 1},
    {.name = "size two below second", .nn = {1, 3}, .k = 2, .want = 3},
    {.name = "size two match second", .nn = {1, 3}, .k = 3, .want = 3},
    // size 3
    {.name = "size three below", .nn = {1, 3, 5}, .k = 0, .want = 1},
    {.name = "size three match first", .nn = {1, 3, 5}, .k = 1, .want = 1},
    {.name = "size three below second", .nn = {1, 3, 5}, .k = 2, .want = 3},
    {.name = "size three match second", .nn = {1, 3, 5}, .k = 3, .want = 3},
    {.name = "size three below third", .nn = {1, 3, 5}, .k = 4, .want = 5},
    {.name = "size three match third", .nn = {1, 3, 5}, .k = 5, .want = 5},
};

INSTANTIATE_TEST_SUITE_P(FindTest, PositiveFindTest,
                         ValuesIn(kPositiveFindTestParams));

inline constexpr int kIllegalValue = 0;  // unused in test - can be any value

const FindTestParam kNegativeFindTestParams[]{
    {.name = "size one above", .nn = {1}, .k = 2, .want = kIllegalValue},
    {.name = "size two above", .nn = {1, 3}, .k = 4, .want = kIllegalValue},
    {.name = "size three above",
     .nn = {1, 3, 5},
     .k = 6,
     .want = kIllegalValue},
};

INSTANTIATE_TEST_SUITE_P(FindTest, NegativeFindTest,
                         ValuesIn(kNegativeFindTestParams));

}  // namespace
}  // namespace iq::lower
